### ISSUE TITLE
TPH Inheritance duplicate relationship error

### ISSUE THREAD

**Issue by prindacerk** [dotnet/efcore#7079]
The logical relationship is that one vehicle can have multiple estimates and invoices. Some estimates are converted as invoice while others aren't.

This is my Code First Classes.
```C#
    public class Vehicle
    {
        public string Id { get; set; }
        public string Variant { get; set; }
        public DateTime LastServiced { get; set; }
        public virtual List<Estimate> Estimates { get; set; }
        public virtual List<Invoice> Invoices { get; set; }
    }

    internal static partial class UserDbContextBuilder
    {
        internal static void ConfigureTable(EntityTypeBuilder<Vehicle> entity)
        {
            entity.ToTable("Vehicles");

            entity.HasKey(vehicle => vehicle.Id);

            entity.Property(vehicle => vehicle.Id)
                .HasMaxLength(IdMaxLength);

            entity.Property(vehicle => vehicle.Variant)
                .HasMaxLength(DefaultStringMaxLength);
        }

        internal static void ConfigureForeignKeys(EntityTypeBuilder<Vehicle> entity)
        {

        }
    }

    public class Estimate
    {
        public string Id { get; set; }
        public string ReferenceNo { get; set; }
        public DateTime CreatedDate { get; set; }
        public DateTime LastUpdated { get; set; }
        public string VehicleId { get; set; }
        public virtual Vehicle Vehicle { get; set; }
    }

    internal static partial class UserDbContextBuilder
    {
        internal static void ConfigureTable(EntityTypeBuilder<Estimate> entity)
        {
            entity.ToTable("Estimates");

            entity.HasKey(estimate => estimate.Id);

            entity.Property(estimate => estimate.Id)
                .HasMaxLength(IdMaxLength);

            entity.Property(estimate => estimate.VehicleId)
                .IsRequired()
                .HasMaxLength(IdMaxLength);

            entity.Property(estimate => estimate.ReferenceNo)
                .HasMaxLength(DefaultStringMaxLength);
        }

        internal static void ConfigureForeignKeys(EntityTypeBuilder<Estimate> entity)
        {
            entity.HasOne(estimate => estimate.Vehicle)
                .WithMany(vehicle => vehicle.Estimates)
                .HasForeignKey(estimate => estimate.VehicleId)
                .OnDelete(DeleteBehavior.Cascade)
                .HasConstraintName("FK_Estimates_Vehicles");
        }
    }

    public class Invoice : Estimate
    {
        public DateTime InvoiceDate { get; set; }
        public DateTime DueDate { get; set; }

    }

    internal static partial class UserDbContextBuilder
    {
        internal static void ConfigureTable(EntityTypeBuilder<Invoice> entity)
        {
        }

        internal static void ConfigureForeignKeys(EntityTypeBuilder<Invoice> entity)
        {
        }
    }
```

When I generate the Migration script, I seem to get two references to the table **Vehicles**. And the script adds a new reference column also. However, when I do a 

> database update

 command, it throws an error stating that the table is referencing the same table twice. I can't execute the script successfully. Can you please help and look at it?

```
System.InvalidOperationException: Cannot create a relationship between 'Vehicle.Invoices' and 'Estimate.Vehicle', because there already is a relationship between 'Vehicle.Estimates' and 'Estimate.Vehicle'. Navigation properties can only participate in a single relationship.
```

The migration script it generated shows as follows.
```C#
            migrationBuilder.CreateTable(
                name: "Vehicles",
                columns: table => new
                {
                    Id = table.Column<string>(maxLength: 40, nullable: false),
                    LastServiced = table.Column<DateTime>(nullable: false),
                    Variant = table.Column<string>(maxLength: 50, nullable: true)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_Vehicles", x => x.Id);
                });

            migrationBuilder.CreateTable(
                name: "Estimates",
                columns: table => new
                {
                    Id = table.Column<string>(maxLength: 40, nullable: false),
                    CreatedDate = table.Column<DateTime>(nullable: false),
                    Discriminator = table.Column<string>(nullable: false),
                    LastUpdated = table.Column<DateTime>(nullable: false),
                    ReferenceNo = table.Column<string>(maxLength: 50, nullable: true),
                    VehicleId = table.Column<string>(maxLength: 40, nullable: false),
                    DueDate = table.Column<DateTime>(nullable: true),
                    InvoiceDate = table.Column<DateTime>(nullable: true),
                    VehicleId1 = table.Column<string>(nullable: true)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_Estimates", x => x.Id);
                    table.ForeignKey(
                        name: "FK_Estimates_Vehicles",
                        column: x => x.VehicleId,
                        principalTable: "Vehicles",
                        principalColumn: "Id",
                        onDelete: ReferentialAction.Cascade);
                    table.ForeignKey(
                        name: "FK_Estimates_Vehicles_VehicleId1",
                        column: x => x.VehicleId1,
                        principalTable: "Vehicles",
                        principalColumn: "Id",
                        onDelete: ReferentialAction.Restrict);
                });
```

Why is it creating a second relationship to the same table instead of filtering the records using the Discriminator column? 
 
### Further technical details

"Microsoft.AspNetCore.Identity": "1.1.0",
"Microsoft.AspNetCore.Identity.EntityFrameworkCore": "1.1.0",
"Microsoft.EntityFrameworkCore": "1.1.0",
"Microsoft.EntityFrameworkCore.Design": "1.1.0",
"Microsoft.EntityFrameworkCore.SqlServer": "1.1.0",
"NETStandard.Library": "1.6.1"

"Microsoft.EntityFrameworkCore.Tools.DotNet": "1.0.0-preview3-final"

**Comment by jnm2** [comment]
(Total side note, which you can feel free to ignore, but inheritance should be a "is-a" relationship, not a "has-a" relationship. In my mind just because an invoice *contains* information from an estimate doesn't mean that an invoice *is* an estimate.)

**Comment by AndriySvyryd** [comment]
@jnm2 Could you also share the DbContext? I tried with this one and didn't get any errors:
```C#

    public class ServicingContext : DbContext
    {
        public DbSet<Vehicle> Vehicles { get; set; }
        public DbSet<Invoice> Invoices { get; set; }
        public DbSet<Estimate> Estimates { get; set; }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            UserDbContextBuilder.ConfigureTable(modelBuilder.Entity<Estimate>());
            UserDbContextBuilder.ConfigureTable(modelBuilder.Entity<Vehicle>());

            UserDbContextBuilder.ConfigureForeignKeys(modelBuilder.Entity<Estimate>());
        }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(
                "Data Source=(localdb)\\MSSQLLocalDB;Database=Servicing;Integrated Security=True;Connect Timeout=30");
        }
    }
```

**Comment by jnm2** [comment]
@AndriySvyryd sorry for interjecting, I think you want @prindacerk

**Comment by AndriySvyryd** [comment]
@jnm2 Oops, you are correct.

**Comment by prindacerk** [comment]
@jnm2 I know and on actual implementation, both Estimate and Invoice have a base class which they inherit from. However, I noticed this problem when implementing the classes and I wanted to bring it to attention that if the base class has a relationship, then the inheriting class also creates a separate relationship by default which causes conflict.

@AndriySvyryd I assume you were asking me. The project compiles and it even creates a migration. But when you try to implement it into the database, it throws the error I have highlighted above. I have given the migration script it generated in the initial post. Can you compare if yours generated the same way?

Anyway, my DBContext is as follows.

```
    public class UserDbContext : IdentityDbContext<ApplicationUser>
    {
        public DbSet<Estimate> Estimates { get; set; }
        public DbSet<Invoice> Invoices { get; set; }

        /// <summary>
        /// This constructor is required to pass the connection details to the base class
        /// </summary>
        /// <param name="options">options containing the context connection details</param>
        public UserDbContext(DbContextOptions<UserDbContext> options) : base(options) { }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);

            UserDbContextBuilder.ConfigureTable(modelBuilder.Entity<Vehicle>());
            UserDbContextBuilder.ConfigureTable(modelBuilder.Entity<Estimate>());
            UserDbContextBuilder.ConfigureTable(modelBuilder.Entity<Invoice>());

            
            UserDbContextBuilder.ConfigureForeignKeys(modelBuilder.Entity<Vehicle>());
            UserDbContextBuilder.ConfigureForeignKeys(modelBuilder.Entity<Estimate>());
            UserDbContextBuilder.ConfigureForeignKeys(modelBuilder.Entity<Invoice>());
        }
    }
```

**Comment by AndriySvyryd** [comment]
@prindacerk Even with the provided context I don't see the error. The migration fragment you supplied is the same that I get.
try adding the `verbose` parameter to see if it gives some useful information:
```
dotnet ef database update --verbose
```

Also even though it's unlikely to be cause this issue, update `Microsoft.EntityFrameworkCore.Tools.DotNet` to `1.1.0-preview4-final`

**Comment by prindacerk** [comment]
@AndriySvyryd Sorry I missed a step in the error messages I was getting. My bad. I just recreated a new project and added the classes. I get the first error message from the script. Included the verbose and this is the output.

```
C:\Users\Developer\Documents\Visual Studio 2015\Projects\EntityCoreTest\src\Autolibrary.Domain.User>dotnet ef database update --verb
ose
Project Autolibrary.Domain.User (.NETFramework,Version=v4.5.1) was previously compiled. Skipping compilation.
Setting app base path C:\Users\Developer\Documents\Visual Studio 2015\Projects\EntityCoreTest\src\Autolibrary.Domain.User\bin\Debug\
net451\win7-x64
Finding DbContext classes...
Using context 'UserDbContext'.
Using database 'EntityCoreTest' on server 'DEV\SQLEXPRESS'.
Opening connection to database 'EntityCoreTest' on server 'DEV\SQLEXPRESS'.
Closing connection to database 'EntityCoreTest' on server 'DEV\SQLEXPRESS'.
Opening connection to database 'EntityCoreTest' on server 'DEV\SQLEXPRESS'.
Executed DbCommand (15ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
SELECT OBJECT_ID(N'__EFMigrationsHistory');
Closing connection to database 'EntityCoreTest' on server 'DEV\SQLEXPRESS'.
Opening connection to database 'EntityCoreTest' on server 'DEV\SQLEXPRESS'.
Closing connection to database 'EntityCoreTest' on server 'DEV\SQLEXPRESS'.
Opening connection to database 'EntityCoreTest' on server 'DEV\SQLEXPRESS'.
Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
SELECT OBJECT_ID(N'__EFMigrationsHistory');
Closing connection to database 'EntityCoreTest' on server 'DEV\SQLEXPRESS'.
Opening connection to database 'EntityCoreTest' on server 'DEV\SQLEXPRESS'.
Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
SELECT [MigrationId], [ProductVersion]
FROM [__EFMigrationsHistory]
ORDER BY [MigrationId];
Closing connection to database 'EntityCoreTest' on server 'DEV\SQLEXPRESS'.
Applying migration '20161121225055_AddVehicle'.
Opening connection to database 'EntityCoreTest' on server 'DEV\SQLEXPRESS'.
Beginning transaction with isolation level 'Unspecified'.
Executed DbCommand (1ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
CREATE TABLE [Vehicles] (
    [Id] nvarchar(40) NOT NULL,
    [LastServiced] datetime2 NOT NULL,
    [Variant] nvarchar(50),
    CONSTRAINT [PK_Vehicles] PRIMARY KEY ([Id])
);
Executed DbCommand (2ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
CREATE TABLE [Estimates] (
    [Id] nvarchar(40) NOT NULL,
    [CreatedDate] datetime2 NOT NULL,
    [Discriminator] nvarchar(max) NOT NULL,
    [LastUpdated] datetime2 NOT NULL,
    [ReferenceNo] nvarchar(50),
    [VehicleId] nvarchar(40) NOT NULL,
    [DueDate] datetime2,
    [InvoiceDate] datetime2,
    [VehicleId1] nvarchar(450),
    CONSTRAINT [PK_Estimates] PRIMARY KEY ([Id]),
    CONSTRAINT [FK_Estimates_Vehicles] FOREIGN KEY ([VehicleId]) REFERENCES [Vehicles] ([Id]) ON DELETE CASCADE,
    CONSTRAINT [FK_Estimates_Vehicles_VehicleId1] FOREIGN KEY ([VehicleId1]) REFERENCES [Vehicles] ([Id]) ON DELETE NO ACTION
);
Closing connection to database 'EntityCoreTest' on server 'DEV\SQLEXPRESS'.
System.Data.SqlClient.SqlException (0x80131904): Column 'Vehicles.Id' is not the same length or scale as referencing column 'Estimat
es.VehicleId1' in foreign key 'FK_Estimates_Vehicles_VehicleId1'. Columns participating in a foreign key relationship must be define
d with the same length and scale.
Could not create constraint. See previous errors.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boole
an asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpl
eResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlCommand.RunExecuteNonQueryTds(String methodName, Boolean async, Int32 timeout, Boolean asyncWrite)
   at System.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, String methodName, Boolean sendToP
ipe, Int32 timeout, Boolean asyncWrite)
   at System.Data.SqlClient.SqlCommand.ExecuteNonQuery()
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, String executeMetho
d, IReadOnlyDictionary`2 parameterValues, Boolean openConnection, Boolean closeConnection)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteNonQuery(IRelationalConnection connection, IReadOnlyDi
ctionary`2 parameterValues, Boolean manageConnection)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationCommandExecutor.ExecuteNonQuery(IEnumerable`1 migrationCommands, IR
elationalConnection connection)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.UpdateDatabase(String targetMigration, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
ClientConnectionId:a46a9c25-6e32-40ca-a2e1-d966ccc21bbd
Error Number:1753,State:0,Class:16
Column 'Vehicles.Id' is not the same length or scale as referencing column 'Estimates.VehicleId1' in foreign key 'FK_Estimates_Vehicles_VehicleId1'. Columns participating in a foreign key relationship must be defined with the same length and scale.
Could not create constraint. See previous errors.
```
Then I decided to manually redeclare the VehicleId in Invoice class which is what caused the above error I mentioned in the OP.

As for `1.1.0-preview4-final`, I upgraded the project but then I kept getting the error 
`Unrecognized option '--config'`
So I downgraded it back to the version I know that worked.

**Comment by AndriySvyryd** [comment]
I see, for this error to be fixed you would need to update to 1.1.0-preview4-final. the --config issue is tracked by https://github.com/aspnet/EntityFramework/issues/7071
As a workaround you can edit the migrations file and add maxLength:
```C#
VehicleId1 = table.Column<string>(maxLength: 40, nullable: true)
```

**Comment by prindacerk** [comment]
@AndriySvyryd ok thanks. I was hoping not to mess with the migration file because then any other person who may remove and add a new migration may not realize the patch work I did. And since I can't use `1.1.0-preview4-final` at the moment, I am left with no choice.

Anyway, I am going with a base class approach so both Estimate and Invoice inherit from that and the link to Vehicle is handled in Estimate & Invoice classes. So this avoids the problem faced here. However, hopefully in the future when I have to inherit a class that has a relationship like I mentioned before, the `1.1.0-preview4-final` would have a fix.
